include/Symbol.h

```diff
@@ -37,7 +38,11 @@ class Symbol {
         kind(kind), name(name), location(location) {}
 
     template<typename T>
-    const T& as() const { return *static_cast<const T*>(this); }
+    const T& as() const {
+        // TODO: if (T::mykind != this->kind) return nullptr;
```

Comment:
It looks like the main issue with this todo is that this returns a const& instead of a pointer, so you can't return something like nullptr. Do we have a reasonable solution in mind?